engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,841 @@
+//   │ SkipAndWaitRedirect        │                              │                                │
+//   │                            │                              │                                │
+//   │                            │                              │ Worker2 Rename Column          │
+//   │                            │                              │◄────────────────────────────   │
+//   │  DDLRedirectRequest        │                              │                                │
+//   │◄───────────────────────────┼──────────────────────────────┤                                │
+//   │────────────────────────────┼─────────────────────────────►│                                │
+//   │  DDLRedirectResponse       │                              │                                │
+//   │                            │                              │                                │
+//   ├────┐                       │                              │                                │
+//   │    │                       │                              │                                │
+//   │    │ Redirect              │                              │                                │
+//   │    │                       │                              │                                │
+//   │◄───┘                       │                              │                                │
+//   │                            │                              │                                │
+
+type tableType int
+
+const (
+	normal tableType = iota
+	conflict
+	final
+)
+
+// TableAgent defines an interface for checkpoint.
+type TableAgent interface {
+	FetchAllDoTables(ctx context.Context, cfg *config.JobCfg) (map[metadata.TargetTable][]metadata.SourceTable, error)
+	FetchTableStmt(ctx context.Context, jobID string, cfg *config.JobCfg, sourceTable metadata.SourceTable) (string, error)
+}
+
+// DDLCoordinator is a coordinator for ddl.
+type DDLCoordinator struct {
+	mu          sync.Mutex
+	tables      map[metadata.TargetTable]map[metadata.SourceTable]struct{}
+	shardGroups map[metadata.TargetTable]*shardGroup
+	pendings    sync.WaitGroup
+	logger      *zap.Logger
+
+	kvClient   metaModel.KVClient
+	tableAgent TableAgent
+	jobID      string
+	jobStore   *metadata.JobStore
+}
+
+// NewDDLCoordinator creates a new DDLCoordinator.
+func NewDDLCoordinator(jobID string, kvClient metaModel.KVClient, tableAgent TableAgent, jobStore *metadata.JobStore, pLogger *zap.Logger) *DDLCoordinator {
+	return &DDLCoordinator{
+		tableAgent:  tableAgent,
+		tables:      make(map[metadata.TargetTable]map[metadata.SourceTable]struct{}),
+		jobID:       jobID,
+		kvClient:    kvClient,
+		shardGroups: make(map[metadata.TargetTable]*shardGroup),
+		jobStore:    jobStore,
+		logger:      pLogger.With(zap.String("component", "ddl_coordinator")),
+	}
+}
+
+// Clear clears metadata.
+func (c *DDLCoordinator) Clear(ctx context.Context) error {
+	c.mu.Lock()
```

Comment:
this locking is in fact not needed because as the lifetime of jobmaster, Coordinate will not be called concurrent with Clear/Reset?